src/main/java/com/github/mizool/jediscache/JedisCache.java

```diff
@@ -0,0 +1,353 @@
+package com.github.mizool.jediscache;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.cache.Cache;
+import javax.cache.configuration.CacheEntryListenerConfiguration;
+import javax.cache.configuration.CompleteConfiguration;
+import javax.cache.configuration.Configuration;
+import javax.cache.configuration.MutableConfiguration;
+import javax.cache.integration.CompletionListener;
+import javax.cache.processor.EntryProcessor;
+import javax.cache.processor.EntryProcessorException;
+import javax.cache.processor.EntryProcessorResult;
+
+import lombok.Data;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+import com.google.common.base.Charsets;
+import com.google.common.collect.ImmutableMap;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.JedisPool;
+import redis.clients.jedis.ScanParams;
+import redis.clients.jedis.ScanResult;
+
+@Slf4j
+@Data
+class JedisCache<K, V> implements Cache<K, V>
+{
+    private final JedisCacheManager cacheManager;
+    private final MutableConfiguration<K, V> configuration;
+    private final JedisPool jedisPool;
+    private final String name;
+    private final JedisCacheKeyConverter<K> keyConverter;
+    private final JedisCacheValueConverter<V> valueConverter;
+    private final byte[] jedisCacheName;
+    private final AtomicBoolean closed;
+
+    public JedisCache(
+        @NonNull JedisCacheManager cacheManager,
+        @NonNull String name,
+        @NonNull Configuration<K, V> configuration,
+        @NonNull JedisPool jedisPool)
+    {
+        this.cacheManager = cacheManager;
+        this.name = name;
+        this.jedisPool = jedisPool;
+
+        //we make a copy of the configuration here so that the provided one
+        //may be changed and/or used independently for other caches. we do this
+        //as we don't know if the provided configuration is mutable
+        if (configuration instanceof CompleteConfiguration)
+        {
+            //support use of CompleteConfiguration
+            this.configuration = new MutableConfiguration<>((CompleteConfiguration<K, V>) configuration);
+        }
+        else
+        {
+            //support use of Basic Configuration
+            MutableConfiguration<K, V> mutableConfiguration = new MutableConfiguration<>();
+            mutableConfiguration.setStoreByValue(configuration.isStoreByValue());
+            mutableConfiguration.setTypes(configuration.getKeyType(), configuration.getValueType());
+            this.configuration = mutableConfiguration;
+        }
+
+        JedisCacheJavaSerializationConverter<K, V>
+            converter
+            = new JedisCacheJavaSerializationConverter<>(configuration.getKeyType(), configuration.getValueType());
+        keyConverter = converter;
```

Comment:
`keyConverter` and `valueConverter` refer to the same object, which implements two interfaces for keys/values yet ends up doing exactly the same thing for both. What's worse is that the `toKey`/`fromValue` method names don't make it obvious whether they serialize or deserialize the value: `fromKey()` actually means "serialize a string key to a byte key".

Let's get rid of the interfaces and use only one variable to refer to the serializer/deserializer, which itself does not care whether it's used for keys or values and thus only exposes `serializer()` and `deserialize()` methods.

Reply:
I oppose the idea of getting rid of the interfaces. The fact that they are implemented by the same class is an implementation detail of said class.
My first instinct was rather to split the Converter into its respective parts but then the code for serialzation/deserialization would have been duped.
If we force the same serializer for both keys and values, future changes might become difficult.

Reply:
Future changes might become slightly more difficult, but that's a cost I'm willing to take here in order to make the **current** code readable. The arguments in my first paragraph effectively say "the code is cryptic and hides more than it reveals". It is a fact that we serialize and deserialize keys and values with the same code, so let's not hide that.